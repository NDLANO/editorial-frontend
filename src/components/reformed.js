/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * A modified version of https://github.com/davezuko/react-reformed (MIT license)
 */

import React from 'react';
import PropTypes from 'prop-types';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { getComponentName } from 'ndla-util';

const makeWrapper = WrappedComponent => {
  class FormWrapper extends React.Component {
    constructor(props, ctx) {
      super(props, ctx);
      this.state = {
        submitted: false,
        model: props.initialModel || {},
        fields: {},
      };
      this.setModel = this.setModel.bind(this);
      this.setProperty = this.setProperty.bind(this);
      this.bindToChangeEvent = this.bindToChangeEvent.bind(this);
      this.bindInput = this.bindInput.bind(this);
      this.setSubmitted = this.setSubmitted.bind(this);
    }

    componentWillReceiveProps(nextProps) {
      const currentModel = this.props.initialModel;
      const nextModel = nextProps.initialModel;

      if (this.props.resetOnInitialModelChange && currentModel !== nextModel) {
        const hasChanges = Object.keys(nextModel).find(
          key => nextModel[key] !== currentModel[key],
        );
        if (hasChanges) {
          this.setModel(nextModel);
        }
      }
    }

    setModel(model) {
      this.setState({ model });
      return model;
    }

    setSubmitted(submitted) {
      this.setState({ submitted });
    }

    setProperty(prop, value, isDirty = true) {
      const model = {
        ...this.state.model,
        [prop]: value,
      };

      if (this.state.fields[prop] && this.state.fields[prop].isDirty) {
        this.setState(prevstate => ({
          model: { ...prevstate.model, [prop]: value },
        }));
      } else {
        this.setState(prevstate => {
          const fields = { ...prevstate.fields, [prop]: { isDirty } };
          return { model: { ...prevstate.model, [prop]: value }, fields };
        });
      }
      return model;
    }

    bindToChangeEvent(e) {
      const { name, type, value } = e.target;
      if (type === 'checkbox') {
        const oldCheckboxValue = this.state.model[name] || [];
        const newCheckboxValue = e.target.checked
          ? oldCheckboxValue.concat(value)
          : oldCheckboxValue.filter(v => v !== value);

        this.setProperty(name, newCheckboxValue);
      } else if (type === 'EditorState') {
        this.setProperty(name, value, value.getCurrentContent().hasText()); // Only set dirty flag if text has changed
      } else if (type === 'SlateEditorState') {
        this.setProperty(name, value); // TODO: Handle dirty flag with SlateEditorState
      } else {
        this.setProperty(name, value);
      }
    }

    bindInput(name) {
      return {
        name,
        value: this.state.model[name] || '',
        onChange: this.bindToChangeEvent,
      };
    }

    render() {
      const nextProps = {
        ...this.props,
        bindInput: this.bindInput,
        bindToChangeEvent: this.bindToChangeEvent,
        model: this.state.model,
        fields: this.state.fields,
        submitted: this.state.submitted,
        setProperty: this.setProperty,
        setSubmitted: this.setSubmitted,
        setModel: this.setModel,
      };

      return React.createElement(WrappedComponent, nextProps);
    }
  }

  FormWrapper.propTypes = {
    initialModel: PropTypes.object, //eslint-disable-line
    resetOnInitialModelChange: PropTypes.bool.isRequired,
  };
  FormWrapper.defaultProps = {
    resetOnInitialModelChange: false,
  };
  FormWrapper.displayName = `Reformed(${getComponentName(WrappedComponent)})`;
  return hoistNonReactStatics(FormWrapper, WrappedComponent);
};

export default makeWrapper;
