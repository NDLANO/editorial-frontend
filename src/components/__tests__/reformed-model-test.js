/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import TestRenderer from 'react-test-renderer';
import reformed from '../reformed';

export const Form = props => {
  const { bindInput, setModel, submitted, setSubmitted } = props;

  const handleSubmit = () => {
    setModel({
      title: '',
      introduction: '',
      content: '',
      image: {
        caption: '',
        alt: '',
      },
    });
    setSubmitted(true);
  };

  return (
    <form onSubmit={handleSubmit}>
      {submitted && <span>Submitted!</span>}
      <input type="text" {...bindInput('title')} />
      <input type="text" {...bindInput('introduction')} />
      <input type="text" {...bindInput('content')} />
      <input type="text" {...bindInput('image.alt')} />
      <input type="text" {...bindInput('image.caption')} />
    </form>
  );
};

Form.propTypes = {
  bindInput: PropTypes.func.isRequired,
  setModel: PropTypes.func.isRequired,
  submitted: PropTypes.bool.isRequired,
  setSubmitted: PropTypes.func.isRequired,
};

const triggerOnChange = (tree, name, value) => {
  const input = tree.children.find(child => child.props.name === name);
  input.props.onChange({ target: { value, name } });
};

const initialModel = {
  title: 'Olé',
  introduction: '',
  content: 'Olé, Olé, Olé',
  image: {
    src: 'image.jpg',
    caption: '',
    alt: 'Alternative image text',
  },
};

test('reformed HOC renderers form component correctly', () => {
  const Reformed = reformed(Form);
  const component = TestRenderer.create(
    <Reformed initialModel={initialModel} />,
  );

  expect(component.toJSON()).toMatchSnapshot();
});

test('reformed HOC handles onChange event correctly', () => {
  const Reformed = reformed(Form);
  const component = TestRenderer.create(
    <Reformed initialModel={initialModel} />,
  );
  const tree = component.toJSON();

  triggerOnChange(tree, 'title', 'Hombre');

  expect(component.toJSON()).toMatchSnapshot();
});

test('reformed HOC handles onChange event on nested property correctly', () => {
  const Reformed = reformed(Form);
  const component = TestRenderer.create(
    <Reformed initialModel={initialModel} />,
  );
  const tree = component.toJSON();

  triggerOnChange(tree, 'image.caption', 'El Grande');

  expect(component.toJSON()).toMatchSnapshot();
});

test('reformed HOC handles setModel and setSubmitted', () => {
  const Reformed = reformed(Form);
  const component = TestRenderer.create(
    <Reformed initialModel={initialModel} />,
  );
  const tree = component.toJSON();

  tree.props.onSubmit();

  expect(component.toJSON()).toMatchSnapshot();
});
