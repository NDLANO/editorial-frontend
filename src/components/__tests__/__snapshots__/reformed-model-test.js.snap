// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reformed HOC handles onChange event correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    name="title"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Hombre"
  />
  <input
    name="introduction"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <input
    name="content"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Olé, Olé, Olé"
  />
  <input
    name="image.alt"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Alternative image text"
  />
  <input
    name="image.caption"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
</form>
`;

exports[`reformed HOC handles onChange event on nested property correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    name="title"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Olé"
  />
  <input
    name="introduction"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <input
    name="content"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Olé, Olé, Olé"
  />
  <input
    name="image.alt"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Alternative image text"
  />
  <input
    name="image.caption"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="El Grande"
  />
</form>
`;

exports[`reformed HOC handles setModel and setSubmitted 1`] = `
<form
  onSubmit={[Function]}
>
  <span>
    Submitted!
  </span>
  <input
    name="title"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <input
    name="introduction"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <input
    name="content"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <input
    name="image.alt"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <input
    name="image.caption"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
</form>
`;

exports[`reformed HOC renderers form component correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    name="title"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Olé"
  />
  <input
    name="introduction"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <input
    name="content"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Olé, Olé, Olé"
  />
  <input
    name="image.alt"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Alternative image text"
  />
  <input
    name="image.caption"
    oBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
</form>
`;
